File: contracts/pixel_index/verify.py
#!/usr/bin/env python3
"""Verify pixel-art-mcp's understanding of pixel-index's custom-asset upload contract
against a live environment.

Consumer-driven, read-only (see checks.py's module docstring for why no authenticated
upload is attempted). A pass means "the zips this repo generates should still be
shape-compatible with what this environment's decode logic expects" — not "this
environment's full API is unchanged."

Run:
  PYTHONPATH=src python -m contracts.pixel_index.verify \\
      --base-url https://pixel-index-api-staging.nntin.xyz
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

import requests

from contracts.pixel_index.checks import CHECKS, CheckResult


def _utc_now() -> str:
    return datetime.now(UTC).isoformat(timespec="seconds").replace("+00:00", "Z")


def run(base_url: str, timeout: float) -> tuple[bool, list[CheckResult]]:
    session = requests.Session()
    context: dict[str, Any] = {}
    results = [check(base_url, session, context, timeout) for check in CHECKS]
    overall_ok = not any(result["status"] == "fail" for result in results)
    return overall_ok, results


def build_result_document(
    env_name: str, base_url: str, ok: bool, results: list[CheckResult]
) -> dict[str, Any]:
    """The stable, machine-readable result consumed by generate_status_site.py."""
    counts = {
        status: sum(result["status"] == status for result in results)
        for status in ("pass", "fail", "skipped")
    }
    failed = [result["name"] for result in results if result["status"] == "fail"]
    return {
        "schema_version": 1,
        "environment": env_name,
        "base_url": base_url,
        "status": "pass" if ok else "fail",
        "checked_at": _utc_now(),
        "counts": counts,
        "checks": results,
        "detail": "" if ok else f"Failed checks: {', '.join(failed)}",
    }


def write_result_document(path: str, document: dict[str, Any]) -> None:
    """Atomically replace a placeholder result with the completed result."""
    destination = Path(path)
    destination.parent.mkdir(parents=True, exist_ok=True)
    temporary = destination.with_name(destination.name + ".tmp")
    temporary.write_text(json.dumps(document, indent=2) + "\n", encoding="utf-8")
    temporary.replace(destination)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--base-url",
        required=True,
        help="pixel-index API base URL, e.g. https://pixel-index-api.nntin.xyz",
    )
    parser.add_argument("--env-name", default=None, help="Label for output, defaults to --base-url")
    parser.add_argument(
        "--output-json",
        default=None,
        help="Write a structured result for the status site.",
    )
    parser.add_argument("--timeout", type=float, default=10.0)
    args = parser.parse_args()

    env_name = args.env_name or args.base_url
    ok, results = run(args.base_url, args.timeout)
    document = build_result_document(env_name, args.base_url, ok, results)

    if args.output_json:
        write_result_document(args.output_json, document)

    lines = [
        f"## pixel-index contract check — {env_name}",
        "",
        "| Check | Result | Detail |",
        "|---|---|---|",
    ]
    icon = {"pass": "✅", "fail": "❌", "skipped": "⚠️"}
    for result in results:
        detail = result["detail"].replace("|", "\\|") or "-"
        status = result["status"]
        lines.append(f"| {result['name']} | {icon[status]} {status} | {detail} |")
    report = "\n".join(lines)

    print(report)
    summary_path = os.environ.get("GITHUB_STEP_SUMMARY")
    if summary_path:
        with open(summary_path, "a", encoding="utf-8") as fh:
            fh.write(report + "\n\n")

    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/contracts-pixel-index/src/verify.ts
#!/usr/bin/env node
/**
 * Verify pixel-art-mcp's understanding of pixel-index's custom-asset upload contract against a
 * live environment. Port of `contracts/pixel_index/verify.py`.
 *
 * Consumer-driven, read-only (see `checks.ts`'s module doc comment for why no authenticated
 * upload is attempted). A pass means "the zips this repo generates should still be
 * shape-compatible with what this environment's decode logic expects" -- not "this environment's
 * full API is unchanged."
 *
 * Run:
 *   node dist/verify.js --base-url https://pixel-index-api-staging.nntin.xyz
 */

import { mkdirSync, renameSync, writeFileSync } from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";

import { CHECKS, type CheckContext, type CheckResult } from "./checks.js";

function utcNow(): string {
  // `datetime.now(UTC).isoformat(timespec="seconds").replace("+00:00", "Z")`: second precision,
  // trailing "Z" rather than "+00:00".
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

export async function run(
  baseUrl: string,
  timeoutSeconds: number,
): Promise<{ ok: boolean; results: CheckResult[] }> {
  const context: CheckContext = {};
  const timeoutMs = timeoutSeconds * 1000;
  const results: CheckResult[] = [];
  for (const check of CHECKS) {
    // Sequential by design (mirrors Python's list comprehension): later checks read `context`
    // fields earlier checks populate (e.g. `checkRoot`'s discovered commit).
    results.push(await check(baseUrl, context, timeoutMs));
  }
  const ok = !results.some((result) => result.status === "fail");
  return { ok, results };
}

export interface ResultDocument {
  readonly schema_version: 1;
  readonly environment: string;
  readonly base_url: string;
  readonly status: "pass" | "fail";
  readonly checked_at: string;
  readonly counts: { readonly pass: number; readonly fail: number; readonly skipped: number };
  readonly checks: readonly CheckResult[];
  readonly detail: string;
}

/** The stable, machine-readable result consumed by `generate-status-site.ts`. */
export function buildResultDocument(
  envName: string,
  baseUrl: string,
  ok: boolean,
  results: readonly CheckResult[],
): ResultDocument {
  const counts = {
    pass: results.filter((result) => result.status === "pass").length,
    fail: results.filter((result) => result.status === "fail").length,
    skipped: results.filter((result) => result.status === "skipped").length,
  };
  const failed = results.filter((result) => result.status === "fail").map((result) => result.name);
  return {
    schema_version: 1,
    environment: envName,
    base_url: baseUrl,
    status: ok ? "pass" : "fail",
    checked_at: utcNow(),
    counts,
    checks: results,
    detail: ok ? "" : `Failed checks: ${failed.join(", ")}`,
  };
}

/** Atomically replace a placeholder result with the completed result. */
export function writeResultDocument(filePath: string, document: ResultDocument): void {
  const destination = path.resolve(filePath);
  mkdirSync(path.dirname(destination), { recursive: true });
  const temporary = `${destination}.tmp`;
  writeFileSync(temporary, JSON.stringify(document, null, 2) + "\n", "utf-8");
  renameSync(temporary, destination);
}

const STATUS_ICON: Record<CheckResult["status"], string> = {
  pass: "✅",
  fail: "❌",
  skipped: "⚠️",
};

export function formatReport(envName: string, results: readonly CheckResult[]): string {
  const lines = [
    `## pixel-index contract check — ${envName}`,
    "",
    "| Check | Result | Detail |",
    "|---|---|---|",
  ];
  for (const result of results) {
    const detail = (result.detail || "-").replaceAll("|", "\\|");
    lines.push(`| ${result.name} | ${STATUS_ICON[result.status]} ${result.status} | ${detail} |`);
  }
  return lines.join("\n");
}

interface Args {
  readonly baseUrl: string;
  readonly envName: string | undefined;
  readonly outputJson: string | undefined;
  readonly timeout: number;
}

function parseArgs(argv: readonly string[]): Args {
  let baseUrl: string | undefined;
  let envName: string | undefined;
  let outputJson: string | undefined;
  let timeout = 10.0;

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === undefined) continue;
    const [flag, inlineValue] = arg.includes("=") ? splitOnce(arg, "=") : [arg, undefined];
    const takeValue = (): string => {
      if (inlineValue !== undefined) return inlineValue;
      i += 1;
      const value = argv[i];
      if (value === undefined) throw new Error(`Missing value for ${flag}`);
      return value;
    };
    switch (flag) {
      case "--base-url":
        baseUrl = takeValue();
        break;
      case "--env-name":
        envName = takeValue();
        break;
      case "--output-json":
        outputJson = takeValue();
        break;
      case "--timeout":
        timeout = Number.parseFloat(takeValue());
        break;
      default:
        throw new Error(`Unknown argument: ${flag}`);
    }
  }
  if (baseUrl === undefined) {
    throw new Error("--base-url is required");
  }
  return { baseUrl, envName, outputJson, timeout };
}

function splitOnce(value: string, separator: string): [string, string] {
  const index = value.indexOf(separator);
  return [value.slice(0, index), value.slice(index + separator.length)];
}

export async function main(argv: readonly string[]): Promise<number> {
  const args = parseArgs(argv);
  const envName = args.envName ?? args.baseUrl;
  const { ok, results } = await run(args.baseUrl, args.timeout);
  const document = buildResultDocument(envName, args.baseUrl, ok, results);

  if (args.outputJson) {
    writeResultDocument(args.outputJson, document);
  }

  const report = formatReport(envName, results);
  console.log(report);
  const summaryPath = process.env["GITHUB_STEP_SUMMARY"];
  if (summaryPath) {
    const { appendFileSync } = await import("node:fs");
    appendFileSync(summaryPath, report + "\n\n", "utf-8");
  }

  return ok ? 0 : 1;
}

/** CLI entrypoint guard: only runs when this module is the process entrypoint
 * (`node dist/verify.js ...`), not when imported by tests. */
const isDirectRun =
  process.argv[1] !== undefined && import.meta.url === pathToFileURL(process.argv[1]).href;
if (isDirectRun) {
  main(process.argv.slice(2))
    .then((code) => {
      process.exitCode = code;
    })
    .catch((error: unknown) => {
      console.error(error instanceof Error ? (error.stack ?? error.message) : String(error));
      process.exitCode = 1;
    });
}
